### PR TITLE
Move logging to new style (remove allocs)

### PR DESCRIPTION
**Pull request by benaadams**
and additionally boxing param issue shown below:

img

Takes allocations for logging calls to zero

### CODE REVIEW

**Comment by rynowak** [comment]
We already have a pattern for this https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/Logging/InnerAttributeRouteLoggerExtensions.cs#L6

**Comment by benaadams** [comment]
That's too weird for me :)

Raised issue #327 for it instead

**Comment by benaadams** [comment]
Actually, quite like it, updated.

Resolves  #327

**Comment by rynowak** [comment]
The pattern for this resolves a few perf issues by doing things like avoiding `params object[]` on both paths, and pre-parsing and caching the format string for you.

If you need to do significant work before calling the delegate, do a manual `IsEnabled` check.

We've measured this and verified that it's allocation free when logging is off, and allocates one object when logging is on. We might take it further in the next milestone to get rid of that allocation when logging is on as well.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelTrace.cs [hunk 1]
```diff
@@ -12,6 +12,32 @@ namespace Microsoft.AspNet.Server.Kestrel
...
+            _connectionStop = LoggerMessage.Define<long>(LogLevel.Debug, 2, @"Connection id ""{ConnectionId}"" stopped.");
+            // ConnectionRead 3
+            _connectionPause = LoggerMessage.Define<long>(LogLevel.Debug, 4, @"Connection id ""{ConnectionId}"" paused.");
+            _connectionResume = LoggerMessage.Define<long>(LogLevel.Debug, 5, @"Connection id ""{ConnectionId}"" resumed.");
+            _connectionReadFin = LoggerMessage.Define<long>(LogLevel.Debug, 6, @"Connection id ""{ConnectionId}"" received FIN.");
+            _connectionWriteFin = LoggerMessage.Define<long>(LogLevel.Debug, 7, @"Connection id ""{ConnectionId}"" sending FIN.");
+            _connectionWroteFin = LoggerMessage.Define<long, int>(LogLevel.Debug, 8, @"Connection id ""{ConnectionId}"" sent FIN with status ""{Status}"".");
+            _connectionKeepAlive = LoggerMessage.Define<long>(LogLevel.Debug, 9, @"Connection id ""{ConnectionId}"" completed keep alive response.");
+            _connectionDisconnect = LoggerMessage.Define<long>(LogLevel.Error, 10, @"Connection id ""{ConnectionId}"" disconnected.");
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelTrace.cs]: Should ApplicationError also use this patern?

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelTrace.cs]: I'd imagine so; but would need the following overload (as `Exception` is is a built in param)?

```csharp
static Action<ILogger, Exception> Define(LogLevel logLevel, int eventId, string formatString)
```
in `Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs` which doesn't currently exist.

**Comment by benaadams** [comment]
Rebased on #337

**Comment by benaadams** [comment]
removed rebase

**Pull request merged**